services/galley/src/Galley/API/MLS/Message.hs

```diff
@@ -1344,43 +1357,63 @@ executeProposalAction loc qusr con (Conv mlsConv) action = do
-      -- FUTUREWORK: update key package ref mapping to reflect conversation membership
-      foldMap
-        ( handleNoChanges
-            . handleMLSProposalFailures @ProposalErrors
-            . fmap pure
-            . updateLocalConversationUnchecked @'ConversationJoinTag lconv qusr con
-            . flip ConversationJoin roleNameWireMember
-        )
-        . nonEmpty
-        . filter (flip Set.notMember (existingMembers lconv))
-        . toList
-
-    removeMembers :: Local Data.Conversation -> NonEmpty (Qualified UserId) -> Sem r [LocalConversationUpdate]
-    removeMembers lconv =
-      foldMap
-        ( handleNoChanges
-            . handleMLSProposalFailures @ProposalErrors
-            . fmap pure
-            . updateLocalConversationUnchecked @'ConversationRemoveMembersTag lconv qusr con
-        )
-        . nonEmpty
-        . filter (flip Set.member (existingMembers lconv))
-        . toList
+existingLocalMembers :: Local Data.Conversation -> Set (Qualified UserId)
+existingLocalMembers lconv =
+  (Set.fromList . map (fmap lmId . tUntagged)) (traverse convLocalMembers lconv)
+
+existingRemoteMembers :: Local Data.Conversation -> Set (Qualified UserId)
+existingRemoteMembers lconv =
+  Set.fromList . map (tUntagged . rmId) . convRemoteMembers . tUnqualified $
+    lconv
+
+existingMembers :: Local Data.Conversation -> Set (Qualified UserId)
+existingMembers lconv = existingLocalMembers lconv <> existingRemoteMembers lconv
+
+addMembers ::
+  HasProposalActionEffects r =>
+  Qualified UserId ->
+  Maybe ConnId ->
+  Local ConvOrSubConv ->
+  NonEmpty (Qualified UserId) ->
+  Sem r [LocalConversationUpdate]
+addMembers qusr con lconvOrSub users = case tUnqualified lconvOrSub of
+  Conv mlsConv -> do
+    let lconv = qualifyAs lconvOrSub (mcConv mlsConv)
+    -- FUTUREWORK: update key package ref mapping to reflect conversation membership
+    foldMap
+      ( handleNoChanges
+          . handleMLSProposalFailures @ProposalErrors
+          . fmap pure
+          . updateLocalConversationUnchecked @'ConversationJoinTag lconv qusr con
+          . flip ConversationJoin roleNameWireMember
+      )
+      . nonEmpty
+      . filter (flip Set.notMember (existingMembers lconv))
+      . toList
+      $ users
+  SubConv _ _ -> pure []
+
+removeMembers ::
+  HasProposalActionEffects r =>
+  Qualified UserId ->
+  Maybe ConnId ->
+  Local ConvOrSubConv ->
+  NonEmpty (Qualified UserId) ->
+  Sem r [LocalConversationUpdate]
+removeMembers qusr con lconvOrSub users = case tUnqualified lconvOrSub of
+  Conv mlsConv -> do
+    let lconv = qualifyAs lconvOrSub (mcConv mlsConv)
+    foldMap
+      ( handleNoChanges
+          . handleMLSProposalFailures @ProposalErrors
+          . fmap pure
+          . updateLocalConversationUnchecked @'ConversationRemoveMembersTag lconv qusr con
+      )
+      . nonEmpty
+      . filter (flip Set.member (existingMembers lconv))
+      . toList
+      $ users
+  SubConv _ _ -> pure []
```

Comment:
Removing is supported, isn't it? I can't find that in the MLS implementation in wire-server page, but I vaguely recall someone giving a use case for this operation.

Reply:
Yes, but there's nothing to do here anyway, because we're not tracking user membership in subconversation (only client membership).